.github/workflows/Publish_npm_packages.yml

```diff
@@ -0,0 +1,139 @@
+name: Nightly npm package publish
+
+on:
+    schedule:
+        - cron: '0 0 * * *' # Runs every night at midnight
+    push:
+        branches:
+            - 'jt/*' # for testing only, remove this line
+    workflow_dispatch: # Allows manual triggering
+
+jobs:
+    publish-npm:
+        runs-on: ubuntu-latest
+
+        steps:
+            - name: Cancel previous runs
+              if: github.event_name != 'schedule'
+              uses: styfle/cancel-workflow-action@0.12.1
+              with:
+                  access_token: ${{ github.token }}
+
+            - name: Checkout code
+              uses: actions/checkout@v4
+
+            - uses: actions/setup-node@v4
+              with:
+                  node-version: '20'
+                  cache: 'yarn'
+            - name: Install node dependencies
+              run: yarn install --immutable
+
+            - name: Authenticate to npm
+              run: echo "//registry.npmjs.org/:_authToken=${{ secrets.NPM_TOKEN }}" > ~/.npmrc
+
+            - name: Check npm login
+              run: |
+                  npm whoami || (echo "Not logged in to npm registry" && exit 1)
+
+            - name: Run Nightly Script
+              env:
+                  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+                  NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
+              run: |
+                  function parse_git_branch() {
+                      git branch 2> /dev/null | sed -e '/^[^*]/d' -e 's/* \(.*\)/\1/'
+                  }
+
+                  function make_pr_description() {
+                      git log origin/main..HEAD
+                  }
+
+                  if [[ "$(git status --porcelain)" != "" ]]; then
+                      echo "There are uncommitted changes. Please commit or stash them before running this script."
+                      exit 1
+                  elif [[ "$(parse_git_branch)" != "main" ]]; then
```

Comment:
You shouldn't fail this, or you should set the tag above to only run this on main.

Reply:
1. Why is this pr incrementing versions and changing build scripts at the same time?
2. Why is all this scripting in the yaml? @sergekh2 didnt we talk about not putting loads is scripting in yaml? Can we move this to a .sh file?
3. We need to pass credentials to lerna publish right? How are you handling that?
4. Nice work

Reply:
1. reverted the commit that increments versions - i was just testing the script in the gh action worfklow
2. we can move this to an .sh file - it was easier to iterate inline at first. Note that the gh action file does need to be in yaml. 
3. yes we are passing creds to npm actually here: https://github.com/river-build/river/pull/379/files#diff-b18da9d93fa6ddfc7c8c164dedbf63e3ec3c240986964eceac8014a8aac6b0caR30. If you have a better way to authenticate, please let me know.
4. 🤝 👍

Reply:
Ah nice. What’s the expiry for that auth? This script takes a while to run.

Reply:
Also this looks like it’s just cut and pasted from my script. Can we just update that one and call it instead. Yes obviously the action needs to be yaml, the reason we put the logic in a bash script is so that anyone can run it anywhere (which also makes it easy to iterate on and debug.)

Reply:
i dont believe the auth expires as the auth token is read from `.npmrc` file. The token itself does expire in 2050. I was not able to create a token in npm w/o an expiry. 

There are a few modifications actually from your script, which i will integrate back in and then also call from the workflow action.